narwhals/_pandas_like/dataframe.py

```diff
@@ -418,7 +419,15 @@ def filter(self, *predicates: IntoPandasLikeExpr, **constraints: Any) -> Self:
             )
             # `[0]` is safe as all_horizontal's expression only returns a single column
             mask = expr._call(self)[0]
+            if len(mask) != len(self):
+                msg = (
+                    f"Predicate result has length {len(mask)}, which is incompatible with "
+                    f"DataFrame of length {len(self)}"
+                )
+                raise ShapeError(msg)
+
             _mask = validate_dataframe_comparand(self._native_frame.index, mask)
```

Comment:
i think this error should be handled within `validate_dataframe_comparand`, perhaps we could add a `preserve_lhs_length` boolean argument to it?
could we then do the equivalent thing for _arrow?

Reply:
Just to make you sure I get this suggestion correctly, you would prefer to add an argument that prevents broadcasting in  `validate_dataframe_comparand`. E.g.

```
def validate_dataframe_comparand(..., allow_broadcast: bool):
    ...
    if len(other) == 1 and not allow_broadcast:
         raise ShapeError(...)
```

Reply:
yup! but only if it's a "length n vs length 1" case, so e.g.
```python
if len(other) ==1 and len(df) > 1 and not allow_broadcast
```
If they're both length-1, then there should be an error

separately to this, we should also do something similar for the series broadcasting cases, so we could raise on user errors in `zip_with` / `when/then/otherwise`

Reply:
Thinking a little more about this while changing `validate_dataframe_comparand`: there would be cases that the shapes are just different, but not necessarily a scalar. I don't know why someone might want to do this but:

```py
df_nw.filter(nw.col("x").unique() > 1)
```

`nw.col("x").unique() > 1` would result to something different than a single boolean, but still with a shape mismatch

Reply:
yup, agree! we could raise for that too in `validate_dataframe_comparand`, right?
- if `allow_broadcast`, `len(other)` can be `len(df)` or `1`
- else, `len(other)` must be `len(df)`